src/main/java/org/spongepowered/common/mixin/core/world/WorldMixin.java

```diff
@@ -595,11 +595,15 @@ public IBlockState getBlockState(final BlockPos pos) {
             return;
         }
         if (this.isTileMarkedForRemoval(pos) && !this.bridge$isFake()) {
+            // We must always return the queued TE for mods
+            // See https://github.com/SpongePowered/SpongeForge/issues/3001
             if (PhaseTracker.getInstance().getCurrentState().allowsGettingQueuedRemovedTiles()) {
                 cir.setReturnValue(this.getQueuedRemovedTileFromProxy(pos));
-                return;
+            } else {
+                // If we're not unwinding, then a mod wants what the tile entity
+                // will be
+                cir.setReturnValue(this.getProcessingTileFromProxy(pos));
```

Comment:
My only reservation is that this should already be failing through to the world's existing set of tile entities.. But it may be a catch 22 case that we *should* be passing back the tile entity from the proxy..

Reply:
>  But it may be a catch 22 case that we should be passing back the tile entity from the proxy..

This was the original problem and why I suggested this as the fix. As far as a mod is concerned, they've done their action, so the previous (now) behaviour returns the current tile entity - the old TE that a mod doesn't expect. 

From memory, TFC does their action, changes the TE and then retrieves it again. In Forge, the TE is the new one, everything is okay for them. In Sponge, the TE is the old one, and thus not as they expect the world to be. This change gives them the new TE, and allows them to continue working as expected (in theory), with their future actions also being captured.